src/outputs/history.cpp

```diff
@@ -45,72 +45,116 @@ namespace parthenon {
-      Kokkos::fence();
-      // apply separate chosen operations to each user-defined history output
-      MPI_Op usr_op;
-      switch (hist_var.hst_op) {
-      case UserHistoryOperation::sum:
-        usr_op = MPI_SUM;
-        break;
-      case UserHistoryOperation::max:
-        usr_op = MPI_MAX;
-        break;
-      case UserHistoryOperation::min:
-        usr_op = MPI_MIN;
-        break;
-      }
-      if (Globals::my_rank == 0) {
-        PARTHENON_MPI_CHECK(MPI_Reduce(MPI_IN_PLACE, &result, 1, MPI_PARTHENON_REAL,
-                                       usr_op, 0, MPI_COMM_WORLD));
-      } else {
-        PARTHENON_MPI_CHECK(MPI_Reduce(&result, &result, 1, MPI_PARTHENON_REAL, usr_op, 0,
-                                       MPI_COMM_WORLD));
+      results[hist_var.hst_op].push_back(result);
+      labels[hist_var.hst_op].push_back(hist_var.label);
+    }
+
+    // Check if the package has enrolled vector history functions which are stored in the
+    // Params under the `hist_vec_param_key` name.
+    if (!params.hasKey(hist_vec_param_key)) {
+      continue;
+    }
+    const auto &hist_vecs = params.Get<HstVec_list>(hist_vec_param_key);
+    for (const auto &hist_vec : hist_vecs) {
+      auto result = hist_vec.hst_vec_fun(md_base.get());
+      for (int n = 0; n < result.size(); n++) {
+        std::string label = hist_vec.label + std::to_string(n);
+        results[hist_vec.hst_op].push_back(result[n]);
+        labels[hist_vec.hst_op].push_back(label);
       }
-#endif
+    }
+  }
 
-      all_results.emplace_back(result);
-      all_labels.emplace_back(hist_var.label);
+  // Need fence so result is ready prior to MPI call
+  Kokkos::fence();
+
+#ifdef MPI_PARALLEL
+  for (auto &op : ops) {
+    if (results[op].size() == 0) {
+      continue;
+    }
+
+    MPI_Op usr_op;
+    if (op == UserHistoryOperation::sum) {
+      usr_op = MPI_SUM;
+    } else if (op == UserHistoryOperation::max) {
+      usr_op = MPI_MAX;
+    } else if (op == UserHistoryOperation::min) {
+      usr_op = MPI_MIN;
+    }
+
+    if (Globals::my_rank == 0) {
+      PARTHENON_MPI_CHECK(MPI_Reduce(MPI_IN_PLACE, &(results[op])[0], results[op].size(),
+                                     MPI_PARTHENON_REAL, usr_op, 0, MPI_COMM_WORLD));
+    } else {
+      PARTHENON_MPI_CHECK(MPI_Reduce(&(results[op])[0], &(results[op])[0],
+                                     results[op].size(), MPI_PARTHENON_REAL, usr_op, 0,
+                                     MPI_COMM_WORLD));
     }
   }
+#endif // MPI_PARALLEL
 
-  // only the master rank writes the file
-  // create filename: "file_basename" + ".hst".  There is no file number.
+  // Only the master rank writes the file.
+  // Create filename: "file_basename.out" + [output number] + ".hst". There is no file
+  // number.
   if (Globals::my_rank == 0) {
     std::string fname;
     fname.assign(output_params.file_basename);
+    fname.append(".out" + std::to_string(output_params.block_number));
```

Comment:
E.g. `advection.hst` is now something like `advection.out3.hst` which is changed behavior.

Reply:
It might be better to make this something machine readable like `%02d` and no `out` something like that?

Reply:
I switched this to use the existing `output_params.file_id` that we use for `.outN.` for other output types like restart and HDF5. I think it is good to keep the `out` because it unifies the output block numbering which is now further being used for histories from the input files.

Reply:
ok that makes sense. Would still prefer 0-padding for the output number.

Reply:
I switched to using the existing `outN` string calculated here: https://github.com/parthenon-hpc-lab/parthenon/blob/cd96dd78141d0d09a957c671f16e78941069e6fe/src/outputs/outputs.cpp#L138

I think if we want to add zero-padding to that, that should be a separate PR/discussion since it could break everybody's file readers.

Reply:
kk works for me.